File: leetcode/Python/permutations.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        sol, visited, permutation = [], [0 for i in range(len(nums))], []
        def _dfs(cur):
            if cur==len(nums):
                sol.append(permutation[:])
            for i in range(len(nums)):
                if visited[i]==1:
                    continue
                visited[i]=1
                permutation.append(nums[i])
                _dfs(cur+1)
                visited[i]=0
                permutation.pop()
        _dfs(0)
        return sol

File: leetcode/C++/permutations.cpp
class Solution {
public:
    vector<int> visited,permutation;
    vector<vector<int>> permutations;
    void solve(int cur, vector<int>& nums){
        if(cur==nums.size()) {
            permutations.push_back(permutation);
            return;
        }
        for(int i=0; i<nums.size(); ++i){
            if(visited[i]) continue;
            visited[i]=1; permutation.push_back(nums[i]);
            solve(cur+1,nums);
            visited[i]=0; permutation.pop_back();
        }
    }
    vector<vector<int>> permute(vector<int>& nums) {
        visited=vector<int>(nums.size(),0);
        solve(0,nums);
        return permutations;
    }
};